### PR TITLE
Add retryable to requests done by MicrosoftAPISession class

### PR DESCRIPTION
**Pull request by artem-shelkovnikov**
## Part of https://github.com/elastic/enterprise-search-team/issues/4701

This change adds retry to Sharepoint Online requests. Retry policy is simple - try 3 times with linear backoff and then bail.

Retries are added to:
- Fetching bearer tokens for Graph API
- Fetching bearer tokens for Sharepoint REST API
- All requests to Graph/REST APIs

Also squeezed minor changes:
- Remove definition of `mock_requests` from test_sharepoint_online.py - already present in ConfTest
- Added `patch_sleep` method that makes all sleeps instant for testing purposes.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

### CODE REVIEW

File: connectors/sources/sharepoint_online.py [hunk 1]
```diff
@@ -260,55 +273,53 @@ async def _get_json(self, absolute_url):
...
+            logger.debug(f"Calling Sharepoint Endpoint: {absolute_url}")
+
+            async with self._http_session.get(
+                absolute_url,
+                headers=headers,
+            ) as resp:
+                yield resp
+                return
+        except ClientResponseError as e:
+            if e.status == 429 or e.status == 503:
```
Comment (tarekziade) [inline connectors/sources/sharepoint_online.py:296]: `if e.status in (429, 503)`

Comment (artem-shelkovnikov) [reply connectors/sources/sharepoint_online.py:296]: Will do in follow-up PRs!

**Review by navarone-feekery — DISMISSED** [review]

File: connectors/sources/sharepoint_online.py [hunk 2]
```diff
@@ -266,55 +274,59 @@ async def _get_json(self, absolute_url):
...
+
+            token = await self._api_token.get()
+            headers = {"authorization": f"Bearer {token}"}
+            logger.debug(f"Calling Sharepoint Endpoint: {absolute_url}")
+
+            async with self._http_session.get(
+                absolute_url,
+                headers=headers,
+            ) as resp:
+                print("YIELDING")
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply connectors/sources/sharepoint_online.py]: Ah my bad! Indeed it was for debugging

**Review by navarone-feekery — APPROVED** [review]
LGTM, one question for my own curiosity.

File: tests/sources/test_sharepoint_online.py [hunk 3]
```diff
@@ -160,11 +159,6 @@ async def token(self):
 
         await session.close()
 
-    @pytest_asyncio.fixture
-    async def mock_responses(self):
-        with aioresponses() as m:
-            yield m
-
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply tests/sources/test_sharepoint_online.py:167]: This function is actually already present in `conftest.py` that is used by all tests by default:
https://github.com/elastic/connectors-python/blob/main/tests/conftest.py#L112

[the remark this answers is not in the document]
Comment (navarone-feekery) [reply tests/sources/test_sharepoint_online.py:167]: Cool, TIL! Thanks for the explanation

**Pull request merged**